### DISCUSSION TITLE
Frequency modulation of 2 arbitrary signals

### DISCUSSION THREAD

**Discussion by ryukau** [ryukau/VSTPlugins#29] [Ideas]
This is a reply to https://github.com/ryukau/VSTPlugins/issues/28#issuecomment-1453630690 .

---

I feel like it's not possible without some heavy compromise. In order to modulate frequency of incoming signal, one has to know the state of phase and waveform of source signal generator. I'm not aware of an algorithm to detect such a thing in real-time.

On the other hand, if you are okay with delay time modulation instead of proper FM, then I can make one. The sound will be something similar to FM on CombDistortion.

**Comment by KlashKontakt** [comment]
there is something similar in the kilohertz company.  phase distortion.  this is an interesting implementation, but in the process I realized that for a greater effect, an input controlled filter for the side signal is needed, since artifacts at high frequencies can give an unpredictable effect.

https://kilohearts.com/products/phase_distortion

**Comment by ryukau** [comment]
Does Phase Distortion introduce latency? If this is the case, it might be using FFT.

**Comment by KlashKontakt** [comment]
Nope! Working fast and low latency.

**Comment by ryukau** [comment]
Okay, then the kind of FM you are talking about isn't in my skill set for now.

Or maybe it's not FM. The description says it's "similar to feedback FM". It might be a phaser that is using input signal or side chain signal to modulate all-pass cutoff frequencies, instead of usual sine wave LFO.

I'll experiment the idea later, and will provide prototype here when I finish implementation. However, please be aware that what I'm doing here is just a hobby project, and I won't make any promise about when it happens.

That said, thanks for the information! Hopefully I find nice tuning sooner.

**Comment by KlashKontakt** [comment]
Of course, I understand and am very glad that you generally agreed to this experiment.  I like your instruments, they have something fresh and informal.  Thanks again for your time and looking forward to the prototype.  I plan to conduct an experiment in renoise by mixing 2 waves in the instrument section for the beginning.

Cheers

**Comment by ryukau** [comment]
@KlashKontakt I implemented the phaser idea. A quick demo video is linked below.  Be careful that the audio on the video below is loud.

https://drive.google.com/file/d/1FEcWpw2lO8voYSU7-5MqIaId8EG1df15/view?usp=sharing

Is this the kind of sound you are looking for? Note that almost all the input signal in the video is sawtooth. It might sounds a bit too simple.

I think this one is fun. So even if it's not close to Phase Distortion, I'll probably release as another plugin.

Just in case, current implementation is lacking following things.

- Side chain modulation.
- Note event to all-pass cutoff.
- Higher ratio oversampling option.
- Separation of DC removal high-pass and feedback high-pass.
- "Too much feedback" notification.

I'll upload binary when above are implemented.

**Comment by KlashKontakt** [comment]
Cool prototype!
well, by the way, this tool has great potential, I really like what happened at 3 minutes of the video

that's what I meant in the new context of fm synthesis
simple demonstration
https://youtu.be/MjMac5jW7QQ

**Comment by ryukau** [comment]
Thanks for the video. I watched official YouTube video, but Phase Distortion sounds more amazing on your video. I haven't expected that it sounds that close to real FM.

I made another attempt, and I guess I found the algorithm. Below is a quick demo video.

https://drive.google.com/file/d/1hisgkosBiHpqnLIdX7iOLtlFXlMVqHBd/view?usp=sharing

Plugin for macOS is available on the link below. Note that the plugin is currently just a proof of concept, and doesn't have side chain and other nice things. Do not save your project with it.

https://drive.google.com/file/d/1rfhhbhagllopRNf2MpGHNcifWBQ2U-KT/view?usp=sharing

I hope this is close enough to Phase Distortion.

---

Below is a bit of technical details.

The idea is to put input signal as a complex number, then rotate the phase as usual FM. Below is the core part of algorithm.

```c++
phase += modulation * input;
phase -= std::floor(phase); // Wrap around to keep phase in [0, 1).
output = input * cos(twopi * phase);
```

Phase rotation is done at last line. It's simplified form of `x * exp(1i * y)`, where `x` and `y` is real number. (Reference: [Rotation matrix](https://en.wikipedia.org/wiki/Rotation_matrix).) The code looks like AM, but accumulation of phase makes it sounds like FM (or I'd call it DX7 style PM).

**Comment by KlashKontakt** [comment]
I tried it this morning and I want to say it's a great tool and even in this video it's very useful in different combinations. The only thing I would add in addition to the low-pass filter is that it still makes sense to add a bend-pass filter. as this can be used to emphasize certain frequencies within the instrument. such as drums. a little later I will show a couple of visual demos to understand the areas of application. Thanks for your work

**Comment by KlashKontakt** [comment]
some experience  and test wit sound design

https://youtu.be/2fFaJ6Fz5z8

**Comment by ryukau** [comment]
Thanks for the video. Just in case, it's better to render the sound before installing the new one linked below. Updating the plugin will break your project.

The plugin ComplexRotation is mostly finished. Download link is below. Side chain is implemented. When highpass frequency is high, `Envelope` can be used to smooth out modulation signal.

- https://drive.google.com/file/d/1vdJyKazly5AxGGA6J6YvQTYL8sW_T2so/view?usp=sharing

Feedback and suggestions are welcome, but keep in mind that I can't guarantee to implement the idea. Also it might be better to put some other name. I prefer descriptive name, but current one is too dry and generic.

**Comment by ryukau** [comment]
For reference, core part of DSP code is linked below.

https://github.com/ryukau/VSTPlugins/blob/3a987999e300d76f2b30c8044ee74c1b3d527355/ComplexRotation/source/dsp/dspcore.cpp#L168-L238

**Comment by KlashKontakt** [comment]
https://youtu.be/UyAHj__Mvk0

test today.
thnx for new version! 

ran a series of tests this version works better. but what was expected as a side chain does not yet work for an outside connection. Or I misunderstood the message. in general, the instrument works well, the only significant point is that the sound changes when changing the sample rate 

<img width="507" alt="Снимок экрана 2023-03-06 в 22 09 44" src="https://user-images.githubusercontent.com/87890552/223208188-39820e14-04a8-4ed5-9c4c-d4698c8e165c.png">

thanks again for your work! even in this form, both plugins look grandiose.

**Comment by ryukau** [comment]
On side chain, make sure to increase `Modulation` in Side Chain section. See following video.

- https://drive.google.com/file/d/1BcGrGA2YF7Qfry7qbVVTAYjPHBCGgpaS/view?usp=sharing

`Modulation` is set to 0 in default. This is to avoid duplication of modulation. FL uses input buffer as side chain buffer when side chain is not connected.

On the difference of sampling rate, I'm not sure if it's my mistake or limitation of DSP. From 2:00 onward on video, I can hear that 96000Hz one is brighter than 44100Hz. However, the change of brightness is sometimes related to interpolation. If possible, please check spectrogram of both signal, and if lower part of frequencies are mostly the same, then it's working as intended. Also the code is available for review.

**Comment by ryukau** [comment]
Ah, the sampling rate issue is my mistake. Following part is not correct. Phase modulation amount must be normalized independent of sampling rate.

https://github.com/ryukau/VSTPlugins/blob/3a987999e300d76f2b30c8044ee74c1b3d527355/ComplexRotation/source/dsp/dspcore.cpp#L82

**Comment by ryukau** [comment]
New version (0.0.2) that patched the sampling rate issue is available on the link below.

- https://drive.google.com/file/d/1lCFotB2zOmKHxD1sdec21C6WtPd74e9b/view?usp=sharing

**Comment by KlashKontakt** [comment]
thanks for your promptness. in ren the side chain is not working and the application itself shows that this connection is mi
<img width="680" alt="Снимок экрана 2023-03-07 в 08 20 11" src="https://user-images.githubusercontent.com/87890552/223328885-e63ef98f-cb57-47aa-931e-0d3e5e2938f1.png">
ssing

<img width="493" alt="Снимок экрана 2023-03-07 в 08 29 50" src="https://user-images.githubusercontent.com/87890552/223329048-f21fe5eb-d44f-4ae6-b016-b05a50014d13.png">

**Comment by ryukau** [comment]
Thanks for the screenshot. I fixed side chain issue, and it worked on Windows. Below is download link.

- https://drive.google.com/file/d/1eAJ2kXvZxRqW7lca27s74_eA9p_6W0IA/view?usp=sharing

Let me know if you still have some issues.

**Comment by KlashKontakt** [comment]
I want to say my first impressions it works! in general, results can be achieved, but there is an element of unpredictability, since it is difficult to track the depth of the effect. and as an option, I tried to bind to the keytracker, but this method also does not always work well, although it may make sense to tinker. in general, I like it, but sometimes I want the modulation to depend more on the height of the input signal than on the value of the parameter.

https://youtu.be/tA2HaVX9fXM

Do you mind if I bring more people to test this plugin?

**Comment by KlashKontakt** [comment]
if you need can i upload renoise projects for study? 

Thanks for Your work!

**Comment by ryukau** [comment]
Okay, I'll add proper gate to ComplexRotation.

I consider here is an open public place. Anyone can join discussion if it's on topic, that is something related to plugin development.

Thanks for the offer, but I'm not familiar with Renoise, so the project won't help much.

**Reply by KlashKontakt** [reply]
Thank you very much!

**Comment by ryukau** [comment]
Gate is added on v0.0.4. Below is download link.

- https://drive.google.com/file/d/1lli-76aEo8_e8GJm9vFW0ru0su7ZN680/view?usp=sharing

`Warp` parameter is also added. CPU load increases when the value is not 0. Sampling rate may change the sound of `Warp` a bit. I guess that's probably caused by rounding error.

**Comment by KlashKontakt** [comment]
I like it more and more instruments start to sound more interesting.  Thank you.
It seems that while everything works, I continue to test.  and from the fact that it would be possible to add a blur effect to slightly smear the sound at the top and reduce its sharpness.

**Comment by ryukau** [comment]
I'm not sure if I understand your comment correctly. Perhaps you are using machine translation, and it's not working as you intended.

The harsh high artifact is inevitable because that's what modulation does. First thing to do is reduce the modulation amount. Second is to use external equalizer with steep gain curve. The filters on ComplexRotation aren't sharp enough to eliminate high. Equalizer can be inserted before or after ComplexRotation.

My understanding is that blurring or smearing is usually achieved by low-pass filter. In other words, blurring is a result of low-pass. Same thing applies to image processing.

---

Edited to fix a word. commend -> comment.

**Comment by KlashKontakt** [comment]
Sorry for my english, yes i use machine translation. Unfortunately, the lack of live practice often puts me in an awkward position. Besides, I'm a little weird myself. =)

I meant a slightly different blur effect. Roughly speaking, this is an effect that is used in reverbs to create depth in a room.

https://youtu.be/ARECireogcs

**Comment by ryukau** [comment]
Thanks for the explanation. I understand the point.

My observation is that reverb blurs spectrogram of the sound, and the blur is usually stronger on lower frequencies. This is mostly because low-pass on feedback path or a side effect of delay interpolation.

To put this other way, reverb must be added to the plugin to get the effect. I won't do it on ComplexRotation, because I'd like to focus high oversampling ratio for ComplexRotation. Reverb costs too much CPU load for 16-fold oversampling. But the idea sounds nice, and I was also thinking about similar idea. That is a feedback comb or FDN, but using the modulation to get FM like effect. I might try the idea in another time, but not for now.

**Comment by ryukau** [comment]
I released the idea discussed here as 2 plugins, AccumulativeRingMod and FeedbackPhaser. Download link is below.

https://github.com/ryukau/VSTPlugins/releases/tag/UhhyouPlugins0.54.0

I close this discussion as completed. If there's something to add, feel free to add comment or open another discussion.

---

Thank you @KlashKontakt for participation. It was a fun little project.
